### PR TITLE
[staging]: texlive: 2017 -> 2018

### LINKED ISSUE #40232: texlive: 2017 -> 2018,  document update process
TeX Live 2018 is out.

As is not obvious how to do update texlive using the scripts in `pkgs/tools/typesetting/tex/texlive`,
a short README documenting the update process would be helpful.

cc maintainers @vcunat @lovek323 @7c6f434c @jwiegley and @veprbl who did the last bump.

### PR DESCRIPTION
**Pull request by xeji**
###### Motivation for this change

Get Texlive 2018 out before the end of 2018, and hopefully into 18.09. Close #40232.

Temporarily based on master for easier building and testing. Will  rebase on staging when ready.

###### Current status

- [x] All of `texlive.bin` builds locally on x64_64-linux
- [x] packages updated to snapshot as of 2018-08-22,
- [x] `texlive.combined.scheme-full` builds
- [x] `pdflatex`/`lualatex` tested with some beamer slides and documents, ok so far
- [ ] The entire toolchain needs a lot of testing - help appreciated! 
- [ ] We need a solution for hosting a stable snapshot of the package sources because official CTAN mirrors are a moving target, not under version control. Possibly IPFS, project-owned servers, ...

---

cc @bhipple @vcunat @veprbl

### CODE REVIEW

File: pkgs/tools/typesetting/tex/texlive/bin.nix [hunk 1]
```diff
@@ -24,16 +24,16 @@ let
...
-      (fetchurl {
-        name = "texlive-poppler-0.59.patch";
-        url = https://git.archlinux.org/svntogit/packages.git/plain/trunk/texlive-poppler-0.59.patch?h=packages/texlive-bin&id=6308ec39bce2a4d735f6ff8a4e94473748d7b450;
-        sha256 = "1c4ikq4kxw48bi3i33bzpabrjvbk01fwjr2lz20gkc9kv8l0bg3n";
-      })
+       (fetchurl {
+         name = "texlive-poppler-0.64.patch";
+         url = https://git.archlinux.org/svntogit/packages.git/plain/trunk/texlive-poppler-0.64.patch?h=packages/texlive-bin;
+         sha256 = "0443d074zl3c5raba8jyhavish706arjcd80ibb84zwnwck4ai0w";
+       })
```
Comment (dotlambda) [inline pkgs/tools/typesetting/tex/texlive/bin.nix]: Why did you add the extra whitespace?

Comment (xeji) [reply pkgs/tools/typesetting/tex/texlive/bin.nix]: That wasn't intentional, maybe my editor's autoindent. Will clean it up before this is merged.

**Comment by dtzWill** [comment]
Hmm, tried building a collection of papers and slides and ran into this:

```
fixed-output derivation produced path '/nix/store/jjmbfglqfz0xlz96dphvi3as4k7i6dfb-ESIEEcv.tar.xz' with sha512 hash '11b0d4anf4r3xawpdcw89v2xf6b91qkmra9z4nhv26p49ra9pqwgnmviksz92dwhyhg6sd9slcsqq2h1hapk51z5548apj9lyf8grvs' instead of the expected hash '0qlm8xnxxpmr3lbbw0jxqsxs5k8n057d10vx2zj0rm8gw43vg0r2np8rm4ah2wx979v52jip4kp4pi6qjglfdrpq88psclirb1dlxkl'
fixed-output derivation produced path '/nix/store/vrch5hq89igspipan0hmf8vgb7ca3zmf-a5comb.tar.xz' with sha512 hash '347chqnshq7vz89nypaa27b6l9g071rgmc50l9f914hdg4fbw3nhyg7v51rndadl322crgvfjd22wpv38p6yc7gywg7cf6ifb2kwm7c' instead of the expected hash '3yicpqj0jicsxp8i0macyjwjmq4l64rwg0saqcj6sbdpvvrps088hmagd6rk1j68rmfp17366bxx6z0d1b3wamb9yi1hsyk3dybndv9'
```

(but only after pulling down many many packages first)

I seem to recall hashes change "quickly" (mentioned somewhere in these expressions)--is that what's going on here, or a missed hash? (or? hehe).

Thanks for working on this, I can at least help test a bit! ^_^

**Comment by dtzWill** [comment]
These packages are being grabbed from the 2017 mirror, I think we need a new one?

**Comment by xeji** [comment]
@dtzwill it doesn't work yet because I haven't pushed the update for the package download URLs yet

**Comment by dtzWill** [comment]
> @dtzWill it doesn't work yet because I haven't pushed the update for the package download URLs yet

Oh, okay! In that case I guess I'd like to report everything is working as expected! ;)

LMK when some testing would be useful, sorry for the noise!

**Comment by xeji** [comment]
...  just pushed the missing commit. `texlive.combined.scheme-full` builds with this and I tested it with some of my beamer slides and papers, seems to work so far.  So now would be the perfect time for some testing :smile:

File: pkgs/tools/typesetting/tex/texlive/bin.nix [hunk 2]
```diff
@@ -24,15 +24,15 @@ let
       # FIXME: a proper mirror, though tarballs.nixos.org saves this case ATM
```
Comment (bhipple) [inline pkgs/tools/typesetting/tex/texlive/bin.nix]: IIRC tarballs.nixos.org does not add everything automatically to the content-address source store; a maintainer has to run this:
https://github.com/NixOS/nixpkgs/blob/master/maintainers/scripts/copy-tarballs.pl

Finding broken/moved tarballs should be easier now, because we're using upstream's `sha512` hashes as published on CTAN instead of adding the layer of indirection to download -> verify -> re-hash with `sha1`; but nevertheless it'd be nice to have the ability to fetch the missing source tarballs from Hydra!

**Comment by bhipple** [comment]
Thanks for carrying the torch on this one @xeji!  

I'm having some issues compiling `texlive.bin`, though.  Are you building with the sandboxes?

```
In file included from ../../../texk/web2c/luatexdir/luasocket/src/auxiliar.c:8:0:
../../../texk/web2c/luatexdir/luasocket/src/auxiliar.h:32:10: fatal error: lua.h: No such file or directory
 #include "lua.h"
          ^~~~~~~
compilation terminated.
make[2]: *** [Makefile:7258: luatexdir/luasocket/src/liblua53socket_a-auxiliar.o] Error 1
make[2]: Leaving directory '/build/texlive-20180414-source/WorkDir/texk/web2c'
```

IIRC this one isn't too big for the bot to build (plus everyone should be asleep now anyways).  Let's see:

@GrahamcOfBorg build texlive.bin

File: pkgs/tools/typesetting/tex/texlive/bin.nix (hunk 2 above)
Comment (vcunat) [reply pkgs/tools/typesetting/tex/texlive/bin.nix]: IIRC the script doesn't work on texlive, because we don't use pure fetching (we unpack, modify a bit, and fix hash of the result).

**Comment by dotlambda** [comment]
A package has already been updated:
```
$ nix build -f . texlive.combined.scheme-full
fixed-output derivation produced path '/nix/store/fna72gc85agnm30gfsiyqm1vrzf8jj2x-gentombow.tar.xz' with sha512 hash '2p6d5kmqahk6k5bv7rc3qiib74368fybr73m97q5xs4znhfwln2pz4ssz8bf18s1zwfk432qqbj7kfn07m75z8d43ayakxzf1c9x2f5' instead of the expected hash '3v79mfjrdw58dh0k6i4qzyanyagr446nw0ksyjap189vla4jl40fq1kwc2ms1a5g9vh4s2cxcqambvqqcx78l3vn1h3jgdb3v2hpmb6'
cannot build derivation '/nix/store/i718qswzrfl8ycrm69bylncbfqd5f3ll-texlive-gentombow-2018.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/bd94c1rbia4yp26yz0cg4azv3r2hr77n-texlive-combined-full-2018.drv': 1 dependencies couldn't be built
[4 built (1 failed)]
error: build of '/nix/store/bd94c1rbia4yp26yz0cg4azv3r2hr77n-texlive-combined-full-2018.drv' failed
```

**Comment by xeji** [comment]
@dotlambda the last commit puts my private snapshot mirror first on the URL list, so this should work now. But we still need a good solution for a stable mirror of the package sources...

**Comment by xeji** [comment]
`aarch64` failure of `texlive.bin` is unrelated. `xindy` reqires `clisp`, which [is broken on aarch64](https://hydra.nixos.org/job/nixpkgs/trunk/clisp.aarch64-linux/all).

But let's see if `texlive.bin.core` builds on all platforms - this shouldn't take too long:
@GrahamcOfBorg build texlive.bin.core

File: pkgs/tools/typesetting/tex/texlive/bin.nix (hunk 2 above)
Comment (xeji) [reply pkgs/tools/typesetting/tex/texlive/bin.nix]: The sources for the binary in this file (`bin.nix`) should be ok, they are archived by release date and should be stable enough. 

The issue is with the thousands of package source tarballs (see `default.nix`), which are not under version control. For packages, the upstream package database file and our generated `pkgs.nix` now both use unmodified `sha512` hashes of the downloaded `.tar.xz` files - so maybe uploading them to `tarballs.nixos.org` could work now?

**Comment by dtzWill** [comment]
Suggested commit: https://github.com/dtzWill/nixpkgs/commit/f05ce5d744440cc56cff011924843cc3c173c878

Otherwise our fmtutil invocation doesn't do anything (compare to TL2017 behavior on NixOS).

------------



> I'm having some issues compiling texlive.bin, though. Are you building with the sandboxes?

I saw this as well, but when I went to debug it with some prints/changes it went away and now I can't reproduce :(.

**Comment by xeji** [comment]
Great, thanks, will add that later today.  fmtutil without effect was present since the latest texlive 2017 update, see discussion  in #40826.  Cc @timokau 
-- 
Sent from my mobile device

**Comment by timokau** [comment]
Great, thanks for looking into that @dtzWill! Will you backport that change to 2017 @xeji? Or just go to 2018 with that fix included?

**Comment by xeji** [comment]
@timokau I am currently checking this patch with both texlive 2017 and 2018. If it works for both it could go into master.

**Comment by xeji** [comment]
The fix in https://github.com/dtzWill/nixpkgs/commit/f05ce5d744440cc56cff011924843cc3c173c878 applies to both texlive 2017 and 2018, so I picked it to master in a separate PR #45466.
Thanks again @dtzWill !

**Comment by xeji** [comment]
@bhipple @dtzWill  the  `lua` related error you both encountered once building `texlive.bin` may be non-deterministic and caused by parallel building. I don't think it's worth investigating for now (and don't want to turn off parallel builds just yet) but let's keep it in mind.

**Comment by xeji** [comment]
Before the last texlive-2017 update in #40826 , the non-binary TeX packages were built as fixed-output derivations. This property was dropped in #40826 for reasons explained there (mainly maintainability, transparency). @vcunat has voiced [concerns about that and suggested](https://github.com/NixOS/nixpkgs/pull/40826#issuecomment-415139963)  that we go back to fixed-output.

It's not a question directly related to the 2017->2018 update. But if we want to go back to fixed-output we should make it part of this update. 

So I would like to hear people's opinions please.

**Comment by vcunat** [comment]
It's related in the way that, if you do both, you will save one hash regeneration by fixing hashes _after_ updating.

**Comment by bhipple** [comment]
Hmm I didn't realize the fixed part of fixedOutput.  Personally I think the improvement in simplicity and transparency is worth rebuilds -- particularly since the upstream hashes are moving and a re-hash internally makes it very hard to figure out how to get the src after the fact -- but admittedly I don't have much knowledge about Hydra space/time limitations and costs.

For users who aren't using Hydra and are rebuilding everything from source, it's quite nice to have hashes that match upstream's hashes, because if one is missing they can go hunting for the missing src package in a relatively straightforward way.  

That said, I'm ok with whichever approach other maintainers would like to take.

**Comment by vcunat** [comment]
Hydra probably wouldn't be affected significantly, except the extra bandwidth possibly (I actually don't know the CDN pricing, etc.).  From my point of view it was mainly about users whose system updates may be (e.g.) several hundred megabytes larger than before.  And optimization OCD, kind of.

**Comment by hmemcpy** [comment]
Hey, sorry to report, seems that this change started breaking my builds. Unfortunately I currently know nothing about Nix/nixpkgs (this will soon change!).
I build an opensource PDF of Bartosz Milewski's book on Category Theory for Programmers. Some time ago, I got contributed a pull request, enabling a CI build using Nix that pulls in Texlive 2017 and builds the book (see https://github.com/hmemcpy/milewski-ctfp-pdf/blob/master/shell.nix)

Seems that 4 days ago, my builds started failing without seemingly any change. It took me a few moments to see that it was pulling Texlive 2018 instead. I guess some packages I'm using aren't compatible with it yet.

Anything I can do to pull Texlive 2017 instead?

**Comment by xeji** [comment]
@hmemcpy this pull request hasn't even been merged yet, so it most certainly doesn't break anything.

**Comment by xeji** [comment]
If you think there is an issue with texlive in current nixpkgs please open an issue describing how to reproduce it.

**Comment by hmemcpy** [comment]
Hmm, true, I followed it from googling, the timing seemed very suspicious. @xeji, is there any reason why I'd get texlive 2018 pulled? I can link to two separate CI logs of the same commit, one passing, one failing.

In the meantime, I looked at both logs, seems the "ok" one pulls in:
```
$ wget --retry-connrefused --waitretry=1 -O /tmp/nix-install https://nixos.org/nix/install
--2018-08-21 06:30:05--  https://nixos.org/nix/install
...
```
and the "broken" gets
```
$ wget --retry-connrefused --waitretry=1 -O /tmp/nix-install https://nixos.org/releases/nix/nix-2.0.4/install
--2018-08-25 13:39:13--  https://nixos.org/releases/nix/nix-2.0.4/install
```

Anyway, I will attempt to describe it in a separate issue as well. Thanks!

**Comment by xeji** [comment]
@hmemcpy please continue this in a new issue beause it's off-topic.

**Comment by vcunat** [comment]
@hmemcpy: this TeX update was merged recently #40826, so it's more likely that one is related.

**Comment by xeji** [comment]
There seem to be pros and cons for both approaches, but on balance I think the saved space and buid time of fixed-output derivations outweigh the drawbacks. So I decided to make the (non-binary) texlive packages fixed-output derivations again. Also added more detailed upgrade instructions.

**Comment by xeji** [comment]
@GrahamcOfBorg build texlive.bin.core

**Comment by xeji** [comment]
@GrahamcOfBorg build texlive.combined.scheme-basic

**Comment by dtzWill** [comment]
Looks like it's missing `fixedHashes.nix`?

**Comment by xeji** [comment]
No, I think it's the sporadic lua.h error you and @bhipple hit before:
```
In file included from ../../../texk/web2c/luatexdir/luasocket/src/auxiliar.c:8:
../../../texk/web2c/luatexdir/luasocket/src/auxiliar.h:32:10: fatal error: 'lua.h' file not found
#include "lua.h"
         ^~~~~~~
1 error generated.
make[2]: *** [Makefile:7258: luatexdir/luasocket/src/liblua53socket_a-auxiliar.o] Error 1
make[2]: Leaving directory '/private/tmp/nix-build-texlive-core-big.bin-2018.drv-0/texlive-20180414-source/WorkDir/texk/web2c'
make[1]: *** [Makefile:16136: all-recursive] Error 1
make[1]: Leaving directory '/private/tmp/nix-build-texlive-core-big.bin-2018.drv-0/texlive-20180414-source/WorkDir/texk/web2c'
make: *** [Makefile:4121: all] Error 2
builder for '/nix/store/9nwsv1s82gbsgjfxqlv5rdqgwbdg7by1-texlive-core-big.bin-2018.drv' failed with exit code 2
```

**Comment by xeji** [comment]
OK, so this seems reproducible on Darwin. But I'm unable to debug it as I don't have a Mac.

**Comment by dtzWill** [comment]
Hmm when I tried this I was getting problems with missing fixedHashes.nix, will investigate then...

Oh, user error it appears? Was using revision b04015b1679e2082c9521d8cb451362b45ff6c37, using 66ea6c4221ffda6d7acc265090f5337af3a1802d does the trick.  Not sure how I got the wrong revision, but it was all clear from the command-line! O:).

**Comment by xeji** [comment]
> For whatever reason github lists the commits in the wrong order? Oh well, sorry for the noise.

That's because I did some interactive rebasing and reordered the commits, but github apparently ignores that and shows them by date.

**Comment by dtzWill** [comment]
Ah okay. Also this last commit looks like might require/use IFD which means this won't eval on Hydra instances and others by default, and can be gc'd freely. I think it is generally discouraged (forbidden?) in nixpkgs.

**Comment by xeji** [comment]
Nevermind that last commit. I removed it, it was supposed to go to another branch, still an experiment.

**Comment by xeji** [comment]
https://github.com/NixOS/nixpkgs/pull/45432/commits/63cca4143c5dba089a545016b8e51c9a7a61114a is an experiment to see if the darwin build error may be related to parallel building.
@GrahamcOfBorg build texlive.bin.core-big

**Comment by xeji** [comment]
No luck, will remove that commit later. Any help fixing the `texlive.bin.core-big` build on Darwin will be greatly appreciated :smile:  Once that's fixed this should be good to go.

**Comment by dtzWill** [comment]
FWIW I encountered that error repeatedly on a server with 32 cores; built fine locally on my laptop (with handful, and maybe limited to one or two).

I'm not sure if it's just that builder, but thought I'd mention.  Would be interesting to see if anyone else experiences that problem on Linux.

**Comment by xeji** [comment]
Interesting. My initial guess was that it is an issue with parallel builds, but at least on Darwin disabling them didn't help.

**Comment by veprbl** [comment]
I'm able to reproduce the problem on Darwin. May I suggest a following fix:
```
diff --git a/pkgs/tools/typesetting/tex/texlive/bin.nix b/pkgs/tools/typesetting/tex/texlive/bin.nix
index 0797e42b4ee..7e50285c9f9 100644
--- a/pkgs/tools/typesetting/tex/texlive/bin.nix
+++ b/pkgs/tools/typesetting/tex/texlive/bin.nix
@@ -175,8 +175,11 @@ core-big = stdenv.mkDerivation { #TODO: upmendex
     ++ withSystemLibs [ "kpathsea" "ptexenc" "cairo" "harfbuzz" "icu" "graphite2" ]
     ++ map (prog: "--disable-${prog}") # don't build things we already have
       [ "tex" "ptex" "eptex" "uptex" "euptex" "aleph" "pdftex"
-        "web-progs" "synctex" "luajittex" "mfluajit" # luajittex is mostly not needed, see:
+        "web-progs" "synctex"
+        "luatex53" # XXX probably can be removed when TexLive 2019 is out
+        # luajittex is mostly not needed, see:
         # http://tex.stackexchange.com/questions/97999/when-to-use-luajittex-in-favour-of-luatex
+        "luajittex" "mfluajit"
       ];

   configureScript = ":";
@@ -184,7 +187,7 @@ core-big = stdenv.mkDerivation { #TODO: upmendex
   # we use static libtexlua, because it's only used by a single binary
   postConfigure = ''
     mkdir ./WorkDir && cd ./WorkDir
-    for path in libs/{teckit,lua52,lua53} texk/web2c; do
+    for path in libs/{teckit,lua52} texk/web2c; do
       (
         if [[ "$path" =~ "libs/lua5" ]]; then
           extraConfig="--enable-static --disable-shared"
@@ -199,8 +202,6 @@ core-big = stdenv.mkDerivation { #TODO: upmendex
   '';

   preBuild = "cd texk/web2c";
-  CXXFLAGS = "-std=c++11 -Wno-reserved-user-defined-literal"; # TODO: remove once texlive 2018 is out?
-  enableParallelBuilding = !stdenv.isDarwin;

   doCheck = false; # fails
```
There seems to be a problem with lua53 flavour of luatex. There should be a simple workaround of forcing ```make``` call, but we likely won't have users that would require luatex53.

**Comment by xeji** [comment]
Thank you @veprbl. Will push the fix after building locally.

**Comment by xeji** [comment]
@GrahamcOfBorg build texlive.bin.core texlive.bin.core-big

**Comment by xeji** [comment]
Looks like  `xindy` needs an extra compiler flag for aarch64.

**Comment by xeji** [comment]
I think this can be merged now. Thanks everyone for your help.

**Comment by timokau** [comment]
One of the first two commits (`poppler_min: 0.61 -> 0.67` or `texlive.bin: 20170524 -> 20180414`) have broken the zathura build:

```
builder for '/nix/store/bkp6dvnsi6nx1yy26vzl0w9zxn22jshl-zathura-core-0.4.0.drv' failed with exit code 1; last 10 log lines:
  [30/38] Compiling C object 'zathura@sta/zathura_shortcuts.c.o'.
  [31/38] Compiling C object 'zathura@sta/zathura_synctex.c.o'.
  FAILED: zathura@sta/zathura_synctex.c.o 
  /nix/store/lzh9rcw3rjvbaj7n3z4nzsvcz85k2p2m-gcc-wrapper-7.3.0/bin/cc -Izathura@sta -I. -I.. -I../ -Idata -I/nix/store/f6raam3phafpy23l3gak2mvk8gpvbky3-gtk+3-3.22.30-dev/include/gtk-3.0 -I/nix/store/cmp7yyigd8l4ix52dp3k0d1qqxxj46wj-glib-2.56.0-dev/include/glib-2.0 -I/nix/store/hvsnfbafr9il3h6bwhb3l8dq9p9gp4vi-glib-2.56.0/lib/glib-2.0/include -I/nix/store/7a5n5xavg8rx1yill07gxd5nhwijmgj3-cairo-1.15.12-dev/include/cairo -I/nix/store/ay7h083ivv8dcw7l4x0w1xb85nxjhi55-freetype-2.9-dev/include/freetype2 -I/nix/store/ay7h083ivv8dcw7l4x0w1xb85nxjhi55-freetype-2.9-dev/include -I/nix/store/xglnyfivgr55wx765mc84v3msyb1mw88-pango-1.42.4-dev/include/pango-1.0 -I/nix/store/hczwzzhm6fgb690dl2cgcsh64rg9jkvg-gdk-pixbuf-2.36.12-dev/include/gdk-pixbuf-2.0 -I/nix/store/y5rbsw4xj6kcnhjl9qfc6np32pbrgcsn-atk-2.28.1-dev/include/atk-1.0 -I/nix/store/3rgzymnavvn74jh0c1xgqdr7n77lx686-girara-0.3.0/include -I/nix/store/cmp7yyigd8l4ix52dp3k0d1qqxxj46wj-glib-2.56.0-dev/include/gio-unix-2.0/ -I/nix/store/6d2bmjmzsxw3gy7pgaiisqggab77bygz-sqlite-3.24.0-dev/include -I/nix/store/0g0g7iflpq6qx2h1v8ljasx6hwcpmgd3-texlive-bin-2018/include/synctex -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c11 -O3 -fPIC -pthread '-DGETTEXT_PACKAGE="zathura"' '-DLOCALEDIR="/nix/store/rz8c1qgz01q7z6p0ba0vb230mpsjpdkf-zathura-core-0.4.0/share/locale"' '-DZATHURA_PLUGINDIR="/nix/store/rz8c1qgz01q7z6p0ba0vb230mpsjpdkf-zathura-core-0.4.0/lib/zathura"' -D_DEFAULT_SOURCE -DWITH_SQLITE -DWITH_SYNCTEX -DWITH_SYNCTEX1 -DWITH_MAGIC -Wall -Wextra -pedantic -Werror=implicit-function-declaration -Werror=vla -fvisibility=hidden  -MD -MQ 'zathura@sta/zathura_synctex.c.o' -MF 'zathura@sta/zathura_synctex.c.o.d' -o 'zathura@sta/zathura_synctex.c.o' -c ../zathura/synctex.c
  In file included from ../zathura/synctex.c:8:0:
  /nix/store/0g0g7iflpq6qx2h1v8ljasx6hwcpmgd3-texlive-bin-2018/include/synctex/synctex_parser.h:53:10: fatal error: synctex_version.h: No such file or directory
   #include "synctex_version.h"
            ^~~~~~~~~~~~~~~~~~~
  compilation terminated.
  ninja: build stopped: subcommand failed.
```

I can't tell which one because texlive doesn't build with only the first one.

**Comment by xeji** [comment]
@timokau zathura was fixed by #46376 , which is in `staging` (but not `staging-next`).

**Comment by timokau** [comment]
Great, sorry for the noise. Just noticed the build failing on 18.09 but I see you already cherry-picked it to staging-18.09 too.
